packages/plugin-ext/src/plugin/tasks/tasks.ts

```diff
@@ -165,52 +195,67 @@ export class TasksExtImpl implements TasksExt {
     }
 
     $provideTasks(handle: number, token: theia.CancellationToken): Promise<TaskDto[] | undefined> {
-        const adapter = this.adaptersMap.get(handle);
-        if (adapter) {
-            return adapter.provideTasks(token).then(tasks => {
+        const provider = this.providersByHandle.get(handle);
+        let addedExecutions = 0;
+        if (provider) {
+            return Promise.resolve(provider.provideTasks(token)).then(tasks => {
                 if (tasks) {
-                    for (const task of tasks) {
-                        if (task.taskType === 'customExecution') {
-                            task.executionId = this.addCustomExecution(task.callback);
-                            task.callback = undefined;
+                    return tasks.map(task => {
+                        const dto = converter.fromTask(task);
+                        if (dto && CustomExecution.is(task.execution!)) {
+                            dto.executionId = this.addProvidedCustomExecution(task.execution);
+                            addedExecutions++;
                         }
-                    }
+                        return dto;
+                    }).filter(task => !!task).map(task => task!);
+                } else {
+                    return undefined;
                 }
+            }).then(tasks => {
+                console.info(`provideTasks: added ${addedExecutions} executions`);
                 return tasks;
             });
```

Comment:
This might be clearer as an `async` function rather than chained promises.
Same immediately below, though that one's less complex.

Reply:
t.b.h, I'm not a huge fan of async/await in particular when the control flow is complicated. A simple reason for it is when we get stack traces from the compiled javascript and it's a PITA to map it back to the original source. But it also sometimes obscures the true control flow. YMMV, of course.

Reply:
I agree about the stack traces, most definitely. `__awaiter__` `__push__` `__generator__` is not a very helpful stack at all. I think some of that noise should go away with the move to transpile to ES6, since at least some of it is TypeScript paraphrasing promises, though I'm not 100% certain how the traces will look later. I won't insist on `async`, but I will present this evidence:

Google for "callback hell": 95,500 results
Google for "async await hell": 3,690 results

:-)

Reply:
+1 for `async-await` - it's more readable for me :-)